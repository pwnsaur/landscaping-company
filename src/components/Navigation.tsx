import styled from 'styled-components';
import Link from 'next/link';
import Image from 'next/image';
import logoImage from '../../public/logo.png';

const Navigation = () => {
  return (
    <Container>
      <StyledLinkLogo href='/'>
        <Logo src={logoImage} alt='logo' width={120} height={70} />
      </StyledLinkLogo>
      <Nav>
        <StyledLink href='/'>SƒÅkums</StyledLink>
        <StyledLink href='/services'>Pakalpojumi</StyledLink>
        <StyledLink href='/projects'>Projekti</StyledLink>
        <StyledLink href='/about'>Par mums</StyledLink>
        <StyledLink href='/contacts'>Kontakti</StyledLink>
      </Nav>
    </Container>
  );
};

export default Navigation;

const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  padding: 1rem 5px;
  background-color: ${(props) => props.theme.colors.background};
  z-index: 1;
`;

const Logo = styled(Image)`
  height: 100%;
  width: auto;
`;

const Nav = styled.nav`
  display: flex;
  justify-content: space-around;
  width: 100%;
  height: 100%;
  max-width: 768px;

  @media (max-width: 36em) {
    flex-wrap: wrap;
  }
`;

const StyledLink = styled(Link)`
  display: flex;
  align-items: center;
  font-size: 1rem;
  color: black;
  text-transform: uppercase;
  height: 2rem;
  margin: auto 0;

  &:hover {
    color: #696767;
    text-decoration: underline;
  }

  @media (max-width: 36em) {
    font-size: 0.8rem;
    margin: 0.2rem 0;
  }
`;

const StyledLinkLogo = styled(StyledLink)`
  height: 100%;
  margin-bottom: 1rem;

  @media (max-width: 36em) {
    width: 100%;
    justify-content: center;
  }
`;
