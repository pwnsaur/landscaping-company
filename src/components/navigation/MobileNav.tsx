import { MouseEvent } from 'react';
import styled, { keyframes } from 'styled-components';

import NavItems from '@components/navigation/NavItems';

const slideIn = keyframes`
  0% {
    transform: translateY(-100%);
  }
  100% {
    transform: translateY(0);
  }
`;

const slideOut = keyframes`
  0% {
    transform: translateY(0);
    visibility: visible;
  }
  100% {
    transform: translateY(-100%);
    visibility: hidden;
  }
`;

type MobileNavProps = {
  isOpen: boolean;
  isVisible: boolean;
  handleItemClick: (event: MouseEvent) => void;
};

const MobileNav = ({ isOpen, isVisible, handleItemClick }: MobileNavProps) => {
  if (!isVisible) return null;

  return (
    <Container>
      <TransparentContainer onClick={handleItemClick}></TransparentContainer>
      <StyledMobileNav $isOpen={isOpen} data-testid='mobile-nav'>
        <NavItems />
      </StyledMobileNav>
    </Container>
  );
};

export default MobileNav;

const Container = styled.div`
  display: flex;
  align-items: center;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

const TransparentContainer = styled.div`
  flex-grow: 1;
  height: 100%;
  background-color: transparent;
`;

const StyledMobileNav = styled.nav<{ $isOpen: boolean }>`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  width: 100%;
  position: fixed;
  background-color: ${({ theme }) => theme.colors.background};
  height: 100%;
  padding-top: 5rem;
  animation: ${({ $isOpen }) => ($isOpen ? slideIn : slideOut)} 0.3s ease-in-out
    forwards;
  z-index: 5;
`;
