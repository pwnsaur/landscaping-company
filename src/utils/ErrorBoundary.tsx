import React, { ErrorInfo } from 'react';

interface Props {
  children: React.ReactNode;
}

interface State {
  hasError: boolean;
}

class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = { hasError: false };
  }

  static getDerivedStateFromError(_: Error): State {
    return { hasError: true };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.log({ error, errorInfo });
  }

  render() {
    if (this.state.hasError) {
      return (
        <div>
          <h1>Lapu neizdevās ielādēt.</h1>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
