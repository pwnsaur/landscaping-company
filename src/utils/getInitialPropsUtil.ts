import { AppContext, AppInitialProps } from 'next/app';

import { isMobileUserAgent } from './userAgent';

export async function getInitialPropsUtil(
  appContext: AppContext
): Promise<AppInitialProps & { isMobile: boolean }> {
  let pageProps = {};

  if (appContext.Component.getInitialProps) {
    pageProps = await appContext.Component.getInitialProps(appContext.ctx);
  }

  const userAgent = appContext.ctx.req?.headers['user-agent'] || 'rekt';
  const isMobile = isMobileUserAgent(userAgent);

  return { pageProps, isMobile };
}
