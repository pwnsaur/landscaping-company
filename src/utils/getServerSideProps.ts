import { GetServerSideProps } from 'next';

import { isMobileUserAgent } from '@utils/userAgent';

export const getServerSideProps: GetServerSideProps = async ({ req }) => {
  const userAgent = req.headers['user-agent'] || 'about rekt';
  const isMobile = isMobileUserAgent(userAgent);
  return { props: { isMobile } };
};
