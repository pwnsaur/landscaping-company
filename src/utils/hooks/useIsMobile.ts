import useWindowSize from './useWindowSize';

const useIsMobile = (): boolean => {
  const { width, height } = useWindowSize();
  const currentWidth = width || 0;
  const currentHeight = height || 0;
  const isMobile = currentWidth <= 640;

  return isMobile;
};

export default useIsMobile;
