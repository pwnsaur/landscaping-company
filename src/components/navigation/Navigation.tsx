import { motion } from 'framer-motion';
import Image from 'next/image';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useState } from 'react';
import styled from 'styled-components';

import logoImage from '@assets/logo.png';
import DesktopNav from '@components/navigation/DesktopNav';
import HamburgerIcon from '@components/navigation/HamburgerIcon';
import MobileNav from '@components/navigation/MobileNav';
import useIsMobile from '@utils/hooks/useIsMobile';

const Navigation = () => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [prevScrollPos, setPrevScrollPos] = useState(0);
  const [isNavBarVisible, setisNavBarVisible] = useState(true);

  const [startUpwardsScrollPos, setStartUpwardsScrollPos] = useState<
    number | null
  >(null);

  const isMobile = useIsMobile();
  const router = useRouter();

  //mobile menu
  const handleMenuItemClick = useCallback(() => {
    setIsOpen(false);
    setIsVisible(false);
  }, []);

  const toggleMenu = () => {
    setIsOpen(!isOpen);
    setTimeout(
      () => {
        setIsVisible(!isOpen);
      },
      !isOpen ? 0 : 300
    );
  };

  useEffect(() => {
    const handleRouteChange = () => handleMenuItemClick();
    router.events.on('routeChangeStart', handleRouteChange);
    return () => router.events.off('routeChangeStart', handleRouteChange);
  }, [router.events, handleMenuItemClick]);

  useEffect(() => {
    handleMenuItemClick();
  }, [isMobile, handleMenuItemClick]);

  //navbar hide
  useEffect(() => {
    const handleScroll = () => {
      const currentScrollPos = window.scrollY;

      if (currentScrollPos <= 100) {
        setisNavBarVisible(true);
        setStartUpwardsScrollPos(null);
      } else if (prevScrollPos > currentScrollPos) {
        if (startUpwardsScrollPos === null) {
          setStartUpwardsScrollPos(prevScrollPos);
        } else if (startUpwardsScrollPos - currentScrollPos >= 100) {
          setisNavBarVisible(true);
          setStartUpwardsScrollPos(null);
        }
      } else if (prevScrollPos < currentScrollPos) {
        setisNavBarVisible(false);
        setStartUpwardsScrollPos(null);
      }

      setPrevScrollPos(currentScrollPos);
    };

    window.addEventListener('scroll', handleScroll);

    return () => window.removeEventListener('scroll', handleScroll);
  }, [prevScrollPos, startUpwardsScrollPos]);

  //scrollbar compensation
  useEffect(() => {
    if (isOpen) {
      const scrollBarWidth =
        window.innerWidth - document.documentElement.clientWidth;
      document.body.style.overflow = 'hidden';
      document.body.style.paddingRight = `${scrollBarWidth}px`;
    } else {
      document.body.style.overflow = '';
      document.body.style.paddingRight = '';
    }
  }, [isOpen]);

  return (
    <Container
      initial={{ y: 0 }}
      animate={{ y: isNavBarVisible ? 0 : -142 }}
      transition={{ ease: 'easeInOut', duration: 0.3 }}
    >
      <Header>
        {isMobile && <HamburgerIcon isOpen={isOpen} onClick={toggleMenu} />}
        <LinkLogo href='/'>
          <Logo
            src={logoImage}
            alt='logo'
            width={120}
            height={70}
            quality={50}
            priority
          />
        </LinkLogo>
        {isMobile ? (
          <MobileNav
            isVisible={isVisible}
            isOpen={isOpen}
            handleItemClick={handleMenuItemClick}
          />
        ) : (
          <DesktopNav />
        )}
      </Header>
    </Container>
  );
};

export default Navigation;

const Container = styled(motion.div)`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  background-color: ${({ theme }) => theme.colors.background};
  z-index: 3;
  position: sticky;
  top: 0;
`;

const Header = styled.header`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;
  max-width: ${({ theme }) => theme.width.normal};
  ${({ theme }) => theme.isMobile && `flex-direction: row;`};
`;

const Logo = styled(Image)`
  height: 100%;
  width: auto;
`;

const LinkLogo = styled(Link)`
  display: flex;
  height: 100%;
  margin: 1rem auto 1rem;

  ${({ theme }) =>
    theme.isMobile &&
    `
      width: 100%;
      justify-content: end;
      margin-right: 1rem;
  `}
`;
