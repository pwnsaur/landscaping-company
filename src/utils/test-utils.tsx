import { ReactElement, ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import { theme } from '@/styles/theme';
import useIsMobile from '@/utils/hooks/useIsMobile';
import { render, RenderOptions } from '@testing-library/react';

interface TestWrapperProps {
  children: ReactNode;
}

const TestWrapper = ({ children }: TestWrapperProps) => {
  const isMobile = useIsMobile();

  return (
    <ThemeProvider theme={{ ...theme, isMobile }}>{children}</ThemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

// eslint-disable-next-line import/export
export * from '@testing-library/react';

// eslint-disable-next-line import/export
export { customRender as render };
