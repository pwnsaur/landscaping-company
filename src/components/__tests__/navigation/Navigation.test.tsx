import MobileNav from '@/components/navigation/MobileNav';
import Navigation from '@/components/navigation/Navigation';
import useIsMobile from '@/utils/hooks/useIsMobile';
import { fireEvent, render, screen } from '@/utils/test-utils';

jest.mock('../../../utils/hooks/useIsMobile');

const mockUseIsMobile = useIsMobile as jest.Mock;

jest.mock('next/router', () => ({
  useRouter: () => ({
    events: {
      on: jest.fn(),
      off: jest.fn(),
    },
  }),
}));

describe('Navigation component', () => {
  test('matches the snapshot', () => {
    const { asFragment } = render(<Navigation />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders mobile navigation when in mobile view', () => {
    mockUseIsMobile.mockReturnValue(true);
    render(
      <MobileNav isOpen={true} isVisible={true} handleItemClick={() => {}} />
    );

    expect(screen.getByTestId('mobile-nav')).toBeInTheDocument();
  });
});
