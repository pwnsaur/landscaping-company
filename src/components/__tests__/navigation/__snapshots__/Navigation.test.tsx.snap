// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component matches the snapshot 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1rem;
  color: #1a1a1a;
  text-transform: uppercase;
  height: 4rem;
  padding: 0 1rem;
  white-space: nowrap;
  -webkit-transition: color 0.3s ease-in-out;
  transition: color 0.3s ease-in-out;
}

.c3:hover {
  color: #696767;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 100%;
  height: 100%;
  max-width: 48rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  background-color: #ededed;
  z-index: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  max-width: 48rem;
}

<div
    class="c0"
  >
    <header
      class="c1"
    >
      <nav
        class="c2"
      >
        <a
          class="c3"
          href="/"
        >
          SƒÅkums
        </a>
        <a
          class="c3"
          href="/services"
        >
          Pakalpojumi
        </a>
        <a
          class="c3"
          href="/projects"
        >
          Projekti
        </a>
        <a
          class="c3"
          href="/about"
        >
          Par mums
        </a>
        <a
          class="c3"
          href="/contacts"
        >
          Kontakti
        </a>
      </nav>
    </header>
  </div>
</DocumentFragment>
`;
